### PR TITLE
sql/schemachanger: Initial support for generating a corpus

### PR DESCRIPTION
**Pull request by fqazi**
These changes introduce the following:

1. Support for gathering declarative schema changer states (corpus) from logictests to help backwards/mixed version compatibility testing
2. Debug command support for replaying corpuses
3. Optional test for testing a generated corpus when a parameter is specified 

Release justification: Mainly test enhancements that are disabled by default, which will allow us to have an automated workload to test declarative schema changer state between versions.

### CODE REVIEW

File: pkg/sql/schemachanger/corpus/corpus.go [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scplan"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/errors"
+)
+
+// Collector used to collect declarative schema changer into a corpus file.
+type Collector struct {
```
Comment (chengxiong-ruan) [inline pkg/sql/schemachanger/corpus/corpus.go:32]: Nice work! Could you add some commentary on how the collector works from high level? specifically, how schema changer states are cached and merged in memory through the tests, why do we need the file lock, and how do we handle corpus generated from different logic test configs (sorry, for this, I think my question is that would the parent test name be the same across different configs and there'd be conflict on corpus file names on a same test).

**Review by fqazi — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86200)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan)

___
*[`pkg/sql/schemachanger/corpus/corpus.go` line 32 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/86200#-N9_6qzbt6mTVp_X3FJe-r2-32:-N9bubCwCgfrR_3PD1i9:b-10ajz2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f8278d3fa02a436f7c90e9566273c3179ef57994/pkg/sql/schemachanger/corpus/corpus.go#L32)):*
<details><summary><i>Previously, chengxiong-ruan (Chengxiong Ruan) wrote…</i></summary><blockquote>

Nice work! Could you add some commentary on how the collector works from high level? specifically, how schema changer states are cached and merged in memory through the tests, why do we need the file lock, and how do we handle corpus generated from different logic test configs (sorry, for this, I think my question is that would the parent test name be the same across different configs and there'd be conflict on corpus file names on a same test).
</blockquote></details>

Added the following, hope it makes it clearer:

// Collector used to collect declarative schema changer into a corpus file. This
// will structure will help gather schema changer states in memory for tests,
// by providing a BeforeStage callback (see scexec.TestingKnobs) that can be
// injected into a cluster. Additionally, the callback will take into account
// the root test name and subtest name when labelling detected statements, which
// allows us to separate configuration for logic test (for example:
// TestLogic/local/system-3-node-tenant_TestLogic/local/system-3-node-tenant_DROP DATABASE test CASCADE_0).
// Once  all the declarative states from a given set of tests are gathered, the
// UpdateCorpus method will then update the on disk corpus file to merge in the
// new entries replacing existing ones. This structure is designed for concurrent
// use both in memory and on disk, meaning that it can parallel threads adding new
// entries. The  UpdateCorpus method uses a lock file to ensure that only a single
// process at a time is merging entries.

File: pkg/sql/schemachanger/corpus/corpus.go [hunk 2]
```diff
@@ -0,0 +1,307 @@
...
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scop"
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scplan"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/cockroach/pkg/util/retry"
+	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/errors"
+)
+
+// Collector used to collect declarative schema changer into a corpus file. This
```
Comment (chengxiong-ruan) [inline pkg/sql/schemachanger/corpus/corpus.go:31]: nit:`Collector used` -> `Collector is used`

**Review by chengxiong-ruan — APPROVED** [review]
💯

File: pkg/sql/schemachanger/sctest/cumulative.go [hunk 3]
```diff
@@ -291,6 +293,51 @@ func Pause(t *testing.T, dir string, newCluster NewClusterFunc) {
 	cumulativeTest(t, dir, testFunc)
 }
 
+// GenerateSchemaChangeCorpus work in progress stuff..
```
Comment (chengxiong-ruan) [inline pkg/sql/schemachanger/sctest/cumulative.go:296]: nit: looks like comment need to be updated

**Review by fqazi — COMMENTED** [review]
@chengxiong-ruan TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/86200#-:-N9e3nliCDVoOrVRKD37:bnd8xg5)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @chengxiong-ruan)

**Pull request merged**